File: leetcode/word_break/WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, unordered_set<string> &dict) {
        if (s.size() == 0) return false; 
        
        vector<bool> T(s.size() + 1, false);  // T[i]: whether the first i letters can be segmented. 
        T[0] = true; 
        
        for (int i = 1; i <= s.size(); i++) {
            int j = i; 
            
            while (j >= 1) {
                string postfix = s.substr(j - 1, i - j + 1); 
                
                if (dict.find(postfix) != dict.end() && T[j - 1]) {
                    T[i] = true; 
                    break; 
                }
                
                j--;
            }
        }
        
        return T[s.size()]; 
    }
};

File: leetcode4/WordBreak.java
public class Solution {
    public boolean wordBreak(String s, Set<String> dict) {
        int len = s.length();
        boolean[] T = new boolean[len + 1];  // T[i]: whether substring ends with s[i] can be break into words

        if (len == 0) {
            return false;
        }

        T[0] = true;

        for (int i = 1; i < len + 1; i++) {  // i: index to T
            for (int j = i - 1; j >= 0; j--) {
                if (T[j] && dict.contains(s.substring(j, i))) {
                    T[i] = true;
                    break; 
                }
            }
        }

        return T[len];
    }
}
